Added state machine to MatchState for restricting Simulations.

LadyEbony opened the pull request:
### Motivation and Context
- [X] Why is this change required? What problem does it solve?
Timer was playing during countdown and game results could theoretically change after the EndEvent but before stage switch. Furthermore, MatchManager was getting too many variables to represent the match's state. This fix addresses it by having "basically global" MatchStateID that represents the current state which can accessed by any script.
- [ ] If it fixes an open issue, include the text `Closes #1` (where 1 would be the issue number) to your commit message.

### Types of changes
What types of changes does your code introduce? Check all the boxes that apply:
- [X] Bug fix (non-breaking change which fixes an issue)
Timer doesn't decrease during countdown.
- [X] Breaking change (fix or feature that would cause existing functionality to change)
MatchState now contains a StateID which represents the Match's, well, current state. Accessible.
What simulations called in MatchSimulation can be altered based on the MatchStateID.

### Description
- [X] Describe your changes in detail.
The commits and change list kinda explain it all. I'm not a too big of my implementation. Seems a little unclean.

LadyEbony committed:
Added MatchStateID to MatchState.
Removed isPause and isControllerRunning from MatchManager (functionality replaced by MatchStateID).
MatchSimulation simulates certain SimulationComponents based on MatchStateID.

james7132 commented on Assets/Code/src/Runtime/Match/Components/MatchManager.cs:
```diff
@@ -51,16 +48,16 @@ public class MatchManager : MonoBehaviour {
+		
+	 Debug.Log("Running match...");
+	 MatchController.CurrentState.StateID = MatchStateID.InGame;
+	 MatchTask = new TaskCompletionSource<MatchResult>();
```
The style for this project uses spaces, please change the tabs to spaces instead.

james7132 commented on Assets/Code/src/Runtime/Match/Components/MatchPauseController.cs:
```diff
@@ -25,9 +25,9 @@ public class MatchPauseController : MonoBehaviour {
   void Update() {
     if (MatchManager == null || !MatchManager.IsLocal) return;
-    if (MatchManager.IsPaused) {
+    if (MatchManager.MatchController.CurrentState.StateID == MatchStateID.Pause) {
```
MatchController.CurrentState.StateID is being called a lot, might be better if there was property that made it easier to access.

james7132 commented on Assets/Code/src/Runtime/Match/MatchSimulation.cs:
```diff
@@ -39,6 +44,14 @@ public class MatchSimulation : IMatchSimulation {
-}
+  void SimulateRestrict(ref MatchState state, MatchInputContext input, params Type[] restrictions) {
+	 foreach (var SimComponent in SimulationComponents) {
+		if (restrictions.Contains(SimComponent.GetType()))
```
Calling GetType on and comparing against types every simulation (60FPS) can get very expensive and may allocate memory. 

Suggested fix: Create separate IMatchSimulation[] arrays for each MatchStateID, put them into a Dictionary<MatchStateID, IMatchSimulation[]> and retrieve the simulation set via state.StateID, then simulate. Uses marginally more permanent memory for less GC allocation and faster execution.

`simulations[state.StateID].Simulate(state, inputContext)`

james7132 commented on Assets/Code/src/Runtime/Match/Components/MatchManager.cs:
```diff
@@ -72,13 +69,16 @@ public class MatchManager : MonoBehaviour {
-    bool changed = paused != IsPaused;
-    IsPaused = paused;
+
+	 var pauseID = paused ? MatchStateID.Pause : MatchStateID.InGame;
```
Needs an additional check: you can only pause if StateID is InGame, if it's not Paused or InGame, you shouldn't be doing either.

james7132 commented on Assets/Code/src/Runtime/Match/Components/MatchManager.cs:
```diff
@@ -31,9 +28,9 @@ public class MatchManager : MonoBehaviour {
   /// </summary>
   void FixedUpdate() {
-    if (!IsPaused) {
+    if (MatchController.CurrentState.StateID != MatchStateID.Pause)
```
Style: Keep the surrounding curly braces.

james7132 commented on Assets/Code/src/Runtime/Match/MatchState.cs:
```diff
@@ -7,6 +7,8 @@
 
 namespace HouraiTeahouse.FantasyCrescendo.Matches {
 
+public enum MatchStateID { Intro, InGame, Pause, End }
```
Document this and each of the enum values.

Also not sure about the naming of MatchStateID, doesn't seem right, but MatchState is already being used. MatchProgressionStage?

james7132 commented on Assets/Code/src/Runtime/Match/MatchState.cs:
```diff
@@ -17,6 +19,7 @@ public class MatchState : INetworkSerializable {
 
   PlayerState[] playerStates;
   public int PlayerCount { get; private set; }
+  public MatchStateID StateID { get; set; } = MatchStateID.Intro;
```
If it doesn't need to be hidden, for these small data objects like this, it's better to just use a public field.

This also should be serialized/deserialized in Serialize/Deserialize below. If you do not know how to do it, leave a TODO for me and I'll get around to doing it.

james7132 commented on Assets/Code/src/Runtime/Core/GameController.cs:
```diff
@@ -19,19 +19,18 @@ public class MatchController : IMatchController {
+	 Simulation.Simulate(ref state, inputContext);
+	 CurrentState = state;
+
+	 Timestep++;
```
Style: four space indent -> two space indent.

LadyEbony committed:
Addressed requested changes.
Fixed code style.
Changed MatchStateID to MatchProgressionState.
New accessor for IMatchController for CurrentProgressionID which is an ease of access function for CurrentState's StateID.
MatchSimulation creates a Dictionary<MatchProgressionID, IMatchSimulations>.

Fixed annoying warning to MatchCountdownUI.

LadyEbony commented:
Address requested changes.
Check commit for more details. (Forget to mention that I added serialization for MatchProgressionState)

james7132 commented on Assets/Code/src/Runtime/Core/GameController.cs:
```diff
@@ -10,23 +10,27 @@ public class MatchController : IMatchController {
 
   public virtual uint Timestep { get; set; }
   public virtual MatchState CurrentState { get; set; }
+  public virtual MatchProgressionState CurrentProgressionID {
```
I should have clarified: put this on MatchManager, not MatchController.

james7132 commented on Assets/Code/src/Runtime/Match/MatchSimulation.cs:
```diff
@@ -12,25 +11,43 @@ public interface IMatchSimulation : IInitializable<MatchConfig>, ISimulation<Mat
+  // Contains all simulations
   readonly IMatchSimulation[] SimulationComponents;
+  // Contains simulations pertaining to the MatchProgressionState
+  Dictionary<MatchProgressionState, IMatchSimulation[]> SimulationDictionary;
```
No need to add Dictionary to the name, the declaration for it already makes it clear. Simulations is declarative enough.

james7132 commented on Assets/Code/src/Runtime/Match/MatchSimulation.cs:
```diff
@@ -12,25 +11,43 @@ public interface IMatchSimulation : IInitializable<MatchConfig>, ISimulation<Mat
-  }
+
+    SimulationDictionary = new Dictionary<MatchProgressionState, IMatchSimulation[]>();
+    SimulationDictionary.Add(
```
More readable if you use a collection initializer: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/how-to-initialize-a-dictionary-with-a-collection-initializer

james7132 commented on Assets/Code/src/Runtime/Match/Components/MatchPauseController.cs:
```diff
@@ -6,7 +6,7 @@ namespace HouraiTeahouse.FantasyCrescendo {
 public class MatchPauseController : MonoBehaviour {
 
-  public MatchManager MatchManager;
+  public MatchManager MatchManager { get { return MatchManager.Instance; } }
```
Style: Please use expression bodied statement: `MatchManager =>MatchManager.Instance` for get-only one-liner properties.

james7132 commented on Assets/Code/src/Runtime/Match/MatchSimulation.cs:
```diff
@@ -39,6 +56,16 @@ public class MatchSimulation : IMatchSimulation {
 
+  IMatchSimulation[] CreateSimulations(IMatchSimulation[] matchSimulations, params Type[] restrictions) {
+    var Simulation = new List<IMatchSimulation>();
+    foreach (var simComponent in matchSimulations) {
```
This is reducible via Linq to `matchSimulations.Where(sim => restrictions.Contains(sim)).ToArray()`

james7132 reviewed (changes requested):
Clarified one of my previous comments. Mostly style and naming changes otherwise.

LadyEbony committed:
Addressed the need for not garbage code style.

LadyEbony commented:
Addressed code style fixes.
